Add etcd API v3 implementation

MKLeb opened the pull request:
### What does this PR do?
Adds an implementation for etcd API v3.

Notes:  The libraries are completely incompatible out of the box, the names for similar queries are completely different (which, albeit, is expected).  However, etcd v2 and v3 are fundamentally incompatible with each other.  Their structures are different:  v2  has a hierarchical construction, meaning it essentially reads like a file system; v3 has no directories, and thus no hierarchy (by design) - its keys are simply interpreted as strings.

This PR allows the use of both APIs, but there are certainly some improvements that could be made (not done here due to time constraints).  Right now, we have a base superclass sitting underneath both API "adapters".  However, the ideal way to do this would be to have the base construction (and all common code) factored out to the top level, using the adapter pattern underneath for all of the fundamental functionality and ensuring they return consistent output to the higher-level object.  Basically, we want the adapters to only implement the CRUD functionality of etcd (and also the `watch` idea).  Then, the single higher-level object would assume their results follow a consistent pattern and would be able to use common code refactored out to do the rest of the processing.

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/60325 and all sub-issues of it

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

MKLeb committed:
add etcd3-py to CI requirements for python_version >= 3.6

Commit:
Add intial etcd3 implementation and adjust the functional tests

Commit:
Merge remote-tracking branch 'upstream/master' into feature/etcd3

Commit:
adjust etcd test suite to test new v3 impl where applicable

MKLeb committed:
pre-commit fun

Commit:
add documentation for new option

MKLeb committed:
Merge branch 'master' into feature/etcd3

waynew reviewed (changes requested):
I think this is probably mostly good, got a few questions, and also would like to poke at this

Ch3LL commented on requirements/static/ci/common.in:
```diff
@@ -14,6 +14,7 @@ clustershell
 croniter>=0.3.0,!=0.3.22"; sys_platform != 'win32'
 dnspython
 docker
+etcd3-py==0.1.6 ; python_version >= '3.6'
```
This is only a requirement for the tests, since its in the `ci` directory

Ch3LL commented on salt/utils/etcd_util.py:
```diff
@@ -29,54 +30,69 @@
     def __init__(
         self,
         opts,
-        profile=None,
```
We need to properly deprecate a kwarg before removing it.

MKLeb replied:
I actually meant to leave this here, will push a fix shortly.

MKLeb commented on salt/modules/etcd_mod.py:
```diff
@@ -54,7 +62,7 @@ def __virtual__():
     Only return if python-etcd is installed
     """
-    if salt.utils.etcd_util.HAS_LIBS:
+    if salt.utils.etcd_util.HAS_ETCD_V2 or salt.utils.etcd_util.HAS_ETCD_V3:
```
Good suggestion, wouldn't be sure how to implement it though.  This config can be in either one of two places, the global config space, or under a profile (which they can have many of, each with their own setting for the `etcd.require_v2`, and is supplied for each call to `etcd_mod` if applicable).  So, in order to keep these as static globals for the module, I think it might have to stay like this (unless you have an idea here)

waynew replied:
I forget precisely the module - I remember something about profile and then getting config based on the profile if it was set.

Hrm. Seems like something we should have a util for :suspect:

waynew replied:
Okay, well I'm not seeing the example that I thought existed. I'll try checking tomorrow(?)

MKLeb commented on salt/pillar/etcd_pillar.py:
```diff
@@ -63,7 +72,10 @@
     import salt.utils.etcd_util
 
-    HAS_LIBS = True
+    if salt.utils.etcd_util.HAS_ETCD_V2 or salt.utils.etcd_util.HAS_ETCD_V3:
```
`See prev. comment`

MKLeb commented on salt/utils/etcd_util.py:
```diff
@@ -126,15 +130,165 @@ def __init__(
+        """
+        raise NotImplementedError()
+
+    def read(self, key, recursive=False, wait=False, timeout=None, waitIndex=None):
```
For consistency, yes.  In order to follow the signature of the old API we exposed and avoid a deprecation cycle, unfortunately, I don't think so.  I can always throw it in as an extra kwarg, but it would be weird to have both `recursive` and `recurse` in the same function signature I think, although I may be alone on that.

waynew replied:
I think since we're going to go ahead and deprecate the v2 support, we should go ahead and take the time to clean up the API while we're at it.

waynew replied:
also from our discussion: `recurse` appears to be "more Salty" - `grep -r "recurse=" salt | wc -l` vs `grep -r "recursive=" salt | wc -l` - they're about the same number of occurrences, but the more core modules have a tendency to use `recurse`.

And generally, we may decide that it's not really worth keeping a consistent API across v2/v3, especially if the API doesn't make any sense with the way we're producing the values.

MKLeb commented on salt/utils/etcd_util.py:
```diff
@@ -497,11 +532,206 @@ def tree(self, path):
+    """
+    Since etcd3 has no concept of directories, this class leaves some methods unimplemented.
+
+    These are: ls and write_directory.
```
I suppose we could pretend, although your suggestion wouldn't quite work I don't think -- it would also show the values of subdirectories since results would hold _anything_ that started with `path`.  I guess it depends on if we want to force directories upon v3 even when the underlying service doesn't have them.  The decision to not support them was based solely on the intuition that users on v3 would already know that v3 doesn't have directories and the concept of `ls` and `write_directory` doesn't make sense with that in mind.  If you think we should pretend, I know how to implement it and I don't think it would take very long.

waynew replied:
I guess I'm conflicted then 😂 

I *think* that in some things like say cache, we provide directory-like APIs even though directories don't actually exist -- because the localfs is built on the filesystem and that mechanism ends out being exposed. So we just fake it where it doesn't exist.

Maybe that *only* makes sense for the cache and not generally for etcd support. Thoughts?

waynew replied:
As we discussed - yes is forever, no is temporary.

I think it's fine for us to say that *currently* we don't need `write_directory` (I'm also OK on punting `ls`). There are a couple of weird edge cases that write_directory would produce that we don't really want to implement right now. We should just document those to the best of our current knowledge and then leave it for future us or someone else to take on when they need that functionality.

MKLeb commented on salt/utils/etcd_util.py:
```diff
@@ -497,11 +532,206 @@ def tree(self, path):
+            result = self._decode_from_bytes(result)
+            ret["key"] = result.key
+            ret["value"] = result.value
+            ret["mIndex"] = getattr(result, "mod_revision", 0)
```
I'd like to ask the developers of the separate libraries about that as well.  I'm not sure if this discrepancy is strictly due to their naming choices, or whether it is a side-effect of the different API responses from etcd itself.  From what I saw, v2 gives us back `mIndex`, and v3 gives us `mod_revision`, but they mean the same thing (what revision altered this key last).

waynew replied:
My initial impression is that we should unify this under `mod_revision` -- but I also accept the possibility that that would be confusing in a backwards sort-of-way 🤔

MKLeb commented on salt/utils/etcd_util.py:
```diff
@@ -497,11 +532,206 @@ def tree(self, path):
+            return None
+        return self.tree(key)
+
+    def read(self, key, recursive=False, wait=False, timeout=None, waitIndex=None):
```
Strictly because this was the signature beforehand.  I could add another kwarg, but as per my earlier comment with regards to recurse/recursive, I'm not sure what we would want to do here.

waynew replied:
unless `waitIndex` is the casing that someone would demand/expect from translating direct API calls, I'll go with my previous comment on deprecating old gnarly bits since we're already deprecating things

MKLeb commented on salt/utils/etcd_util.py:
```diff
@@ -497,11 +532,206 @@ def tree(self, path):
+                    log.error("etcd3 read: No values found for key %s", key)
+                else:
+                    for kv in kvs:
+                        kv = self._decode_from_bytes(kv)
```
From what I saw here, https://github.com/Revolution1/etcd3-py/search?q=encoding, utf-8 is pretty much hardcoded in the v3 library.

waynew replied:
I see I wasn't terribly clear there. I guess what I mean to say is, consider the following:

```
with open('foo', 'r') as f:
    print(type(f.read()))
    
with open('foo', 'rb') as f:
    print(type(f.read()))
```

You'll get bytes from one and text/str from the other. Another example:

```
>>> msgpack.loads(msgpack.dumps({'foo': 'bar'}), raw=False)
{'foo': 'bar'}
>>> msgpack.loads(msgpack.dumps({'foo': 'bar'}), raw=True)
{b'foo': b'bar'}
```

That can be particularly important because in Python `b'foo' != 'foo'`

```
>>> msgpack.loads(msgpack.dumps({'foo': 'bar', b'foo': 'baz'}), raw=False)
{'foo': 'bar', b'foo': 'baz'}
>>> msgpack.loads(msgpack.dumps({'foo': 'bar', b'foo': 'baz'}), raw=True)
{b'foo': b'baz'}
```

waynew replied:
Oh. No.


Nonononononono.

Unfortunately, as we discovered on our call, etcd v3 API has an _incredible_ API, namely all keys are bytes -- period. Also values(?). So... that's uh... definitely a thing 😬

waynew replied:
Upon further discussion, I think the best approach here is to provide an optional (default true) encode key/values using msgpack. The result of msgpack encoding is a byte string, which will be correctly stored in etcd. The msgpack encoded bytestring can safely round-trip, and unlike the native etcd API, it *also* distinguishes between `b'foo'` and `'foo'`. Which is something that at least in Python definitely makes sense.

Of course, if someone is using etcd in an out-of-band sort of way (perhaps someone is doing `etcdctl put foo somevalue` for instance) then it won't make sense to use msgpack in this case. Instead we'll just have to let the ambiguity play out. But that will be an opt-in behavior -- and they might want to decode the key/value coming out, so they should be able to specify decoding and encoding behavior on a global scale as well as on a per-state or per-module call (which would override the global value, natch).

MKLeb commented on salt/utils/etcd_util.py:
```diff
@@ -497,11 +532,206 @@ def tree(self, path):
+        for k, v in fields.items():
+            if isinstance(v, dict):
+                # Not hard failing here so we don't get a partial update
+                log.debug("etcd3 has no concept of directories, skipping key %s", k)
```
I think so, yeah.

MKLeb commented on salt/utils/etcd_util.py:
```diff
@@ -497,11 +532,206 @@ def tree(self, path):
+    # Figure out which API version they are using...
+    use_v2 = conf.get("etcd.require_v2", True)
+    if use_v2:
+        salt.utils.versions.warn_until(
```
Yay, I like that

MKLeb commented on tests/pytests/functional/modules/test_etcd_mod.py:
```diff
@@ -144,13 +158,14 @@ def test_basic_operations(subtests, profile_name, prefix):
+                    "{}/2/".format(prefix): {},
+                },
+            }
+            assert etcd_mod.ls_(path=prefix, profile=profile_name) == expected
```
Good catch, will do

MKLeb commented on tests/pytests/functional/utils/test_etcd_util.py:
```diff
@@ -161,7 +181,7 @@ def test_read(subtests, etcd_client, prefix):
         result = etcd_client.read("{}/read/1".format(prefix))
         assert result
-        assert result.value == "one"
+        assert (result if use_v2 else result.pop()).value == "one"
```
Yeah, read returns a list no matter if it's one 0 results or 100.  Is that weird?  Yes.  Very weird? Certainly.

Commit:
change warnings per suggestions

Commit:
change names for backwards compatibility and add back profile kwarg

Commit:
pre-commit and lint fixes

MKLeb committed:
Merge branch 'master' into feature/etcd3

Commit:
switch docker logic in etcd sdb integration tests

Commit:
Merge remote-tracking branch 'origin/feature/etcd3' into feature/etcd3

Commit:
fix sdb etcd tests v2

Commit:
add ls pseudo-functionality to etcd api v3

Commit:
move towards recursive kwarg deprecation by adding recurse kwarg as preferred

Ch3LL commented:
@waynew can you re-review this PR please

Commit:
add encoding feature and other etcd client args for v3

MKLeb committed:
precommit

waynew commented on salt/utils/etcd_util.py:
```diff
@@ -202,25 +404,32 @@ def get(self, key, recurse=False):
+        if recursive is not None:
+            salt.utils.versions.warn_until(
+                "Argon",
+                "Starting with the Argon release, the recursive kwarg will be deprecated."
```
**suggestion** s/deprecated/removed. It's a terminology thing that we've struggled with for a long time :sweat_smile: Generally it should read something to the effect:

"The recursive argument has been deprecated, and will be removed in the Argon release. Use `recurse` instead."

We also need a `.deprecated` changelog entry (60325.deprecated) listing everything that we're deprecating from here that will be removed in the Argon release :)

waynew commented on salt/utils/etcd_util.py:
```diff
@@ -202,25 +404,32 @@ def get(self, key, recurse=False):
+            )
+            recurse = recursive
+        if kwargs:
+            log.warning("Invalid kwargs passed in will not be used: %s", kwargs)
```
**suggestion** if checking kwargs here then on line 410 that should be `kwargs.pop('recursive', None)`

waynew reviewed (changes requested):
I haven't checked it out and banged on it a bit, so it's entirely possible that I'm missing something else there, I did catch a couple of things from a visual inspection - otherwise this mostly looks :+1:

Commit:
Merge remote-tracking branch 'upstream/master' into feature/etcd3

Commit:
add deprecations changelog and deprecate waitIndex/index kwargs in favor of start_revision

Commit:
pre-commit and revert etcd sdb tests to master version

Commit:
remove unused import

waynew commented on salt/utils/etcd_util.py:
```diff
@@ -897,8 +962,8 @@ def get_conn(opts, profile=None, **kwargs):
-            "Starting with the Potassium release, etcd API v3 will be the default "
-            "and etcd API v2 will be deprecated.",
+            "etcd API v2 has been deprecated.  It will be removed in "
+            "the Potassium release, and etcd API v3 will be the default.",
```
**comment** technically it won't be the default, but the only option, if v2 is removed :trollface:

waynew commented:
Okay, leaving this mostly as a note to self - all identified code changes look like they've been resolved, I'm checking out now and gonna bang on things a bit :+1:

waynew commented:
okay, this took me a while to actually get setup and running on my new system, but I was able to give it a cursory glance and it's looking good. I want to spend a bit more time banging on it tomorrow before I :+1: it

waynew reviewed (changes requested):
Okay, finally getting to bang on this a bit and I think that something
is off, but I'm not entirely sure if it's my expectation or something else
:joy: 

I'm starting with this docker container:

    docker run --rm -it -e ALLOW_NONE_AUTHENTICATION=yes -e ETCD_ENABLE_V2=true --publish 2379:2379 --publish 2380:2380 bitnami/etcd:3

Then I have the following pillar data:

**top**

```
base:
  'kevin':
    - 61911_etcd
```

**`61911_etcd.sls`**

```
some_etcd_config:
  etcd.host: 127.0.0.1
  etcd.port: 2379
  etcd.require_v2: True
```

And the following states:

**top**

```
base:
  'kevin':
    - 61911/base
```

**61911/base.sls**

```
{% if salt['config.get']('some_etcd_config:etcd.require_v2') %}
make with a directory:
  etcd.directory:
    - name: some cool dirname/with some spaces/and a couple of things deep
    - profile: some_etcd_config
{%- endif %}

basic key set:
  etcd.set:
    - name: so_basic
    - value: vanilla ice
    - profile: some_etcd_config

set this key yo:
  etcd.set:
    - name: some/cool/key
    - value: more keys are keyful

some deleted watch:
  etcd.wait_rm:
    - name: del/taco
    - recurse: True

some set watch:
  etcd.wait_set:
    - name: buh/leeted
    - value: this should only set if watch'd

create a thing lulz:
  etcd.set:
    - name: del/taco/del/mar
    - value: goodbye
    - watch_in:
      - etcd: some deleted watch
      - etcd: some set watch

nope:
  etcd.rm:
    - name: del/taco
    - recurse: True
```

Now, when I have v2 enabled, I can run this with `salt kevin state.apply` and
all works. I can even run a couple of modules as well:


     salt kevin etcd.get so_basic
     salt kevin etcd.get some/cool/key
     salt kevin etcd.get "some cool dirname/with some spaces/and a couple of things deep"
     salt kevin etcd.ls /
     salt kevin etcd.ls /some/cool
     salt kevin etcd.ls /some/cool/key
     salt kevin etcd.tree /some
     salt kevin etcd.tree /some/cool

     salt kevin etcd.rm /some/cool/key
     salt kevin etcd.watch /some/cool/key timeout=5
     salt kevin etcd.watch /some/cool/key  # then re-run the states

     salt kevin etcd.watch /some/cool/key
     # then run this next one in a different window
     salt kevin etcd.update /some/cool/key awesome

And all of those work. Super swell!

However, when I swap to v3:

    sed -i 's/True/False/' srv/pillar/61911_etcd.sls
    docker run --rm -it -e ALLOW_NONE_AUTHENTICATION=yes -e ETCD_ENABLE_V2=false --publish 2379:2379 --publish 2380:2380 bitnami/etcd:3

Then running a `state.apply` fails with a bunch of `etcd.EtcdException: Bad Response : 404 page not found`. So there's definitely something going on here - not sure if it's just some misconfiguration on my part, but we probably shouldn't be showing *that* exception to the user anyway.

I can show you what I did tomorrow if that would be helpful :+1:

Commit:
fix mod_watch disbatching

Commit:
Merge remote-tracking branch 'upstream/master' into feature/etcd3

Commit:
Make sure to warn if no profile is given and we have a possible connection error

Commit:
precommit

waynew committed:
Fix test failures with wrong profile

waynew committed:
Fix v3 -> no v2 error

Previously, when the API was v3-only, but the profile wasn't present
leading to the v2 fallback, it would 404 in certain circumstances.

We can't use a more particular exception in the `_etcd_action` because
the salt loader plays havoc with things.

garethgreenaway committed:
Merge branch 'master' into feature/etcd3
